### ISSUE TITLE
Suggests <algorithm> for std::move (correct is <utility>)

### ISSUE THREAD

**Issue by sesse** [include-what-you-use/include-what-you-use#486]
Hi,

As the title says:

```c++
#include <utility>
#include <algorithm>

int foo(int x)
{
  int k = 3;
  return std::move(k);
}
```

Running this with IWYU removes the wrong header

```
atum17:~> /usr/bin/iwyu -std=c++14 -c test.cc       

test.cc should add these lines:

test.cc should remove these lines:
- #include <utility>  // lines 1-1

The full include-list for test.cc:
#include <algorithm>  // for move
---
```

There _is_ an std::move in <algorithm>, but it's a different overload (`OutputIt move( InputIt first, InputIt last, OutputIt d_first )`). This is probably what's tripping it up.

**Comment by kimgr** [comment]
Can you run IWYU with a higher verbosity level? That might give some clues as to its reasoning. I can't find any direct mapping from `std::move` to `<algorithm>`, so maybe there's a deeper mistake somewhere.

**Comment by sesse** [comment]
The reason why it wants \<algorithm\> seems pretty clear to me; there are two entirely different functions called std::move, and one of them (although the wrong one) is in \<algorithm\>.

Attaching a log with --verbose=10.

[iwyu-log.txt](https://github.com/include-what-you-use/include-what-you-use/files/1362181/iwyu-log.txt)

**Comment by kimgr** [comment]
No, I think there's a mapping bug:

```
Marked full-info use of decl 0x5555575a7468 k (from test.cc:6:7) at test.cc:7:20
test.cc:7:10: (5) [ FunctionCall ] 0x5555575a7d00 template<> constexpr typename std::remove_reference<int &>::type &&move<int &>(int &&__t) noexcept
Marked full-info use of decl 0x55555720d3c0 std::move (from /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/bits/move.h:98:5) at test.cc:7:10
...
Mapped /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/bits/move.h to <algorithm> for std::move (#include already present)
```
iwyu_include_picker.cc has this:

    // This didn't come from the grep, but seems to be where swap()
    // is defined?
    { "<bits/move.h>", kPrivate, "<algorithm>", kPublic },   // for swap<>()

Maybe we should revisit this mapping; it's from the GCC 4.4-4.6 timeframe.

**Comment by jcageman** [comment]
On a side note, the mappings for msvc are also not very up to date. I currently have quite some additional mappings to ensure i get the right includes.

**Comment by kimgr** [comment]
@jcageman -- it's not so much about up-to-date-ness for the MSVC headers. It's more that IWYU only has built-in support for GCC, for historical reasons. We'd like to do something about it, but just haven't found a good, principled solution yet.

**Comment by sesse** [comment]
I'm sure this is incredibly naïve and has been discussed before, but the immediate thing that comes to mind is to make the mapping on symbols instead of files :-) Ie., simply declare that the C++ standard says std::function comes from \<functional\>. That certainly would make some things easier in my own project headers as well, especially when people start adding the same typedef in five different .h files and IWYU picks the wrong one.

**Comment by kimgr** [comment]
@sesse No, I think that makes sense, it just doesn't scale very well. There are many more types than headers in the standard libraries, and whenever a new type is added to an existing header, the mappings would need to be revisited. With the current private-to-public mapping approach, things work out nicely until a new header is added, which should be a much less common event. That said, the current approach doesn't work very well, so we should explore our options. It just seems like so much work, and I have so little time :)

**Comment by sesse** [comment]
You could always try to screenscrape cppreference.com :-)

**Comment by kimgr** [comment]
I actually started to work on that at some point, but lost steam before I got anything useful out of it. Likewise with the standards drafts from GitHub :)

**Comment by bungeman** [comment]
I just ran into this myself, and found this issue. I found the line in the mappings in iwyu_include_picker.cc mentioned above. On my machine bits/move.h defines std::forward, std::move, and std::swap, so there will also be an issue for std::forward (I think it just gets used less that std::move). An additional wrinkle here is that C++11 actually moved std::swap from <algorithm> to <utility>, so for anyone compiling with C++11 the current suggestion is always wrong for everything in this file (even fro std::swap).
